src/surface.cpp

```diff
@@ -192,6 +195,59 @@ Surface::reflect(Position r, Direction u) const
   return u -= (2.0 * projection / magnitude) * n;
 }
 
+Direction
+Surface::diffuse_reflect(Position r, Direction u) const
+{
+  // Diffuse reflect direction according to the normal.
+  // cosine distribution 
+  
+  Direction n = normal(r);
+  n = n/n.norm();
+  const double projection = std::max(-1.0, std::min(1.0, n.dot(u)));
```

Comment:
Given that `projection` is just used to decide whether to make `cosine` positive or negative, it would be ok to just assign
```suggestion
  const double projection = n.dot(u);
```